sparse/_coo/numba_extension.py

```diff
@@ -0,0 +1,272 @@
+
+
+@unbox(COOType)
+def unbox_COO(typ: COOType, obj: COO, c) -> NativeValue:
+    ret_ptr = cgutils.alloca_once(c.builder, c.context.get_value_type(typ))
+    is_error_ptr = cgutils.alloca_once_value(c.builder, cgutils.false_bit)
+    fail_obj = c.context.get_constant_null(typ)
+
+    with ExitStack() as stack:
+        data = _unbox_native_field(typ.data_type, obj, "data", c)
+        then, otherwise = stack.enter_context(c.builder.if_else(data.is_error, likely=False))
+        with then:
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+            c.builder.store(fail_obj, ret_ptr)
+        stack.enter_context(otherwise)
+
+        coords = _unbox_native_field(typ.coords_type, obj, "coords", c)
+        then, otherwise = stack.enter_context(c.builder.if_else(coords.is_error, likely=False))
+        with then:
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+            c.builder.store(fail_obj, ret_ptr)
+        stack.enter_context(otherwise)
+
+        shape = _unbox_native_field(typ.shape_type, obj, "shape", c)
+        then, otherwise = stack.enter_context(c.builder.if_else(shape.is_error, likely=False))
+        with then:
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+            c.builder.store(fail_obj, ret_ptr)
+        stack.enter_context(otherwise)
+
+        fill_value = _unbox_native_field(typ.fill_value_type, obj, "fill_value", c)
+        then, otherwise = stack.enter_context(c.builder.if_else(fill_value.is_error, likely=False))
+        with then:
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+            c.builder.store(fail_obj, ret_ptr)
+        stack.enter_context(otherwise)
+
+        coo = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+        coo.coords = coords.value
+        coo.data = data.value
+        coo.shape = shape.value
+        coo.fill_value = fill_value.value
+        c.builder.store(cgutils.false_bit, is_error_ptr)
+        c.builder.store(coo._getvalue(), ret_ptr)
+    return NativeValue(c.builder.load(ret_ptr), is_error=c.builder.load(is_error_ptr))
+
+
+@box(COOType)
+def box_COO(typ: COOType, val: 'some LLVM thing', c) -> COO:
+    ret_ptr = cgutils.alloca_once(c.builder, c.pyapi.pyobj)
+    fail_obj = c.pyapi.get_null_object()
+
+    # there's no way to early return here - we have no choice but to nest like crazy...
+
+    coo = cgutils.create_struct_proxy(typ)(c.context, c.builder, value=val)
+
+    with ExitStack() as stack:
+        data_obj = c.box(typ.data_type, coo.data)
+        then, otherwise = stack.enter_context(c.builder.if_else(cgutils.is_null(c.builder, data_obj), likely=False))
+        with then:
+            c.builder.store(fail_obj, ret_ptr)
+        stack.enter_context(otherwise)
+        coords_obj = c.box(typ.coords_type, coo.coords)
+
+        then, otherwise = stack.enter_context(c.builder.if_else(cgutils.is_null(c.builder, coords_obj), likely=False))
+        with then:
+            c.pyapi.decref(data_obj)
```

Comment:
Would it be possible to somehow create a context manager that did this `decref` and then remove all these?

Reply:
No, because the thing to remember is that `with then` is not a branch, it always runs every time.

However, we could keep a list that we iterate over the same way in each `then`.

Reply:
> However, we could keep a list that we iterate over the same way in each `then`.

Let's do this then.

Reply:
To wit:
```diff
--- C:\Users\wiese\Repos\forks\sparse\sparse\_coo\numba_extension.py	Mon Jan 20 15:37:12 2020
+++ C:\Users\wiese\Repos\forks\sparse\sparse\_coo\numba_extension.py	Mon Jan 20 15:42:57 2020
@@ -213,36 +213,42 @@
 
     with ExitStack() as stack:
         coo = cgutils.create_struct_proxy(typ)(c.context, c.builder, value=val)
+        to_cleanup = []
 
         data_obj = c.box(typ.data_type, coo.data)
         with early_exit_if_null(c.builder, stack, data_obj):
-            c.builder.store(fail_obj, ret_ptr)
+            for o in to_cleanup[::-1]:
+                c.pyapi.decref(o)
+            c.builder.store(fail_obj, ret_ptr)
+        to_cleanup.append(data_obj)
 
         coords_obj = c.box(typ.coords_type, coo.coords)
         with early_exit_if_null(c.builder, stack, coords_obj):
-            c.pyapi.decref(data_obj)
-            c.builder.store(fail_obj, ret_ptr)
+            for o in to_cleanup[::-1]:
+                c.pyapi.decref(o)
+            c.builder.store(fail_obj, ret_ptr)
+        to_cleanup.append(coords_obj)
 
         shape_obj = c.box(typ.shape_type, coo.shape)
         with early_exit_if_null(c.builder, stack, shape_obj):
-            c.pyapi.decref(coords_obj)
-            c.pyapi.decref(data_obj)
-            c.builder.store(fail_obj, ret_ptr)
+            for o in to_cleanup[::-1]:
+                c.pyapi.decref(o)
+            c.builder.store(fail_obj, ret_ptr)
+        to_cleanup.append(shape_obj)
 
         fill_value_obj = c.box(typ.fill_value_type, coo.fill_value)
         with early_exit_if_null(c.builder, stack, fill_value_obj):
-            c.pyapi.decref(shape_obj)
-            c.pyapi.decref(coords_obj)
-            c.pyapi.decref(data_obj)
-            c.builder.store(fail_obj, ret_ptr)
+            for o in to_cleanup[::-1]:
+                c.pyapi.decref(o)
+            c.builder.store(fail_obj, ret_ptr)
+        to_cleanup.append(fill_value_obj)
 
         class_obj = c.pyapi.unserialize(c.pyapi.serialize_object(COO))
         with early_exit_if_null(c.builder, stack, class_obj):
-            c.pyapi.decref(shape_obj)
-            c.pyapi.decref(coords_obj)
-            c.pyapi.decref(data_obj)
-            c.pyapi.decref(fill_value_obj)
-            c.builder.store(fail_obj, ret_ptr)
+            for o in to_cleanup[::-1]:
+                c.pyapi.decref(o)
+            c.builder.store(fail_obj, ret_ptr)
+        to_cleanup.append(class_obj)
 
         args = c.pyapi.tuple_pack([coords_obj, data_obj, shape_obj])
         c.pyapi.decref(shape_obj)
```
Not sure it's worth it.

Reply:
@hameerabbasi, note that this results in more lines not fewer. Do you still want me to go ahead with this?

Reply:
No, I think this is fine as-is now. Your points are valid, it might be better to build an object holding this data in a follow-up PR.